File: sdk/go/provider/expressinboxhub.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

/* expressinboxhub.com：GET 首页提取 CSRF token，POST /messages 创建邮箱并获取邮件 */

const expressinboxhubBase = "https://expressinboxhub.com"

var expressinboxhubCsrfRe = regexp.MustCompile(`<meta\s+name="csrf-token"\s+content="([^"]+)"`)

/* expressinboxhubBrowserHeaders 设置浏览器模拟请求头 */
func expressinboxhubBrowserHeaders(req *http.Request) {
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8")
	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
	req.Header.Set("Cache-Control", "no-cache")
	req.Header.Set("DNT", "1")
	req.Header.Set("Pragma", "no-cache")
	req.Header.Set("Upgrade-Insecure-Requests", "1")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36")
	req.Header.Set("sec-ch-ua", `"Chromium";v="146", "Not-A.Brand";v="24", "Google Chrome";v="146"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
}

/* expressinboxhubAjaxHeaders 设置 AJAX POST 请求头 */
func expressinboxhubAjaxHeaders(req *http.Request) {
	req.Header.Set("Accept", "application/json, text/plain, */*")
	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8")
	req.Header.Set("Origin", expressinboxhubBase)
	req.Header.Set("Referer", expressinboxhubBase+"/")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36")
	req.Header.Set("X-Requested-With", "XMLHttpRequest")
	req.Header.Set("sec-ch-ua", `"Chromium";v="146", "Not-A.Brand";v="24", "Google Chrome";v="146"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
	req.Header.Set("sec-fetch-dest", "empty")
	req.Header.Set("sec-fetch-mode", "cors")
	req.Header.Set("sec-fetch-site", "same-origin")
}

/* expressinboxhubExtractCSRF 从 HTML 中提取 CSRF token */
func expressinboxhubExtractCSRF(html string) (string, error) {
	if m := expressinboxhubCsrfRe.FindStringSubmatch(html); len(m) > 1 {
		return m[1], nil
	}
	return "", fmt.Errorf("expressinboxhub: 未找到 CSRF token")
}

/* expressinboxhubMergeCookies 合并响应中的 Set-Cookie 为 Cookie 字符串 */
func expressinboxhubMergeCookies(existing string, cookies []*http.Cookie) string {
	parts := []string{}
	if existing != "" {
		parts = append(parts, existing)
	}
	for _, c := range cookies {
		parts = append(parts, c.Name+"="+c.Value)
	}
	return strings.Join(parts, "; ")
}

/*
 * ExpressinboxhubGenerate 创建 expressinboxhub 临时邮箱
 * 流程: GET 首页获取 CSRF token 和 session cookies → POST /messages 创建邮箱
 */
func ExpressinboxhubGenerate() (*CreatedMailbox, error) {
	/* 第一步: GET 首页，提取 CSRF token 并收集 session cookies */
	getReq, err := http.NewRequest("GET", expressinboxhubBase, nil)
	if err != nil {
		return nil, err
	}
	expressinboxhubBrowserHeaders(getReq)

	getResp, err := HTTPClient().Do(getReq)
	if err != nil {
		return nil, err
	}
	defer getResp.Body.Close()
	if err := CheckHTTPStatus(getResp, "expressinboxhub session"); err != nil {
		return nil, err
	}

	cookie := expressinboxhubMergeCookies("", getResp.Cookies())
	raw, err := io.ReadAll(getResp.Body)
	if err != nil {
		return nil, err
	}
	html := string(raw)

	csrf, err := expressinboxhubExtractCSRF(html)
	if err != nil {
		return nil, err
	}

	/* 第二步: POST /messages 携带 CSRF token 创建邮箱 */
	body := "_token=" + csrf
	postReq, err := http.NewRequest("POST", expressinboxhubBase+"/messages", strings.NewReader(body))
	if err != nil {
		return nil, err
	}
	expressinboxhubAjaxHeaders(postReq)
	postReq.Header.Set("Cookie", cookie)
	postReq.Header.Set("X-CSRF-TOKEN", csrf)

	postResp, err := HTTPClient().Do(postReq)
	if err != nil {
		return nil, err
	}
	defer postResp.Body.Close()
	if err := CheckHTTPStatus(postResp, "expressinboxhub generate"); err != nil {
		return nil, err
	}

	/* 合并新的 session cookies */
	cookie = expressinboxhubMergeCookies(cookie, postResp.Cookies())

	postRaw, err := io.ReadAll(postResp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Mailbox  string `json:"mailbox"`
		Messages []any  `json:"messages"`
	}
	if err := json.Unmarshal(postRaw, &result); err != nil {
		return nil, fmt.Errorf("expressinboxhub: 响应解析失败: %s", string(postRaw[:min(len(postRaw), 120)]))
	}

	emailAddr := strings.TrimSpace(result.Mailbox)
	if emailAddr == "" {
		return nil, fmt.Errorf("expressinboxhub: 响应中未包含邮箱地址")
	}

	/* token 中编码 cookie 和 CSRF，后续获取邮件时需要 */
	token := cookie + "\n" + csrf

	return &CreatedMailbox{
		Channel: "expressinboxhub",
		Email:   emailAddr,
		Token:   token,
	}, nil
}

/*
 * ExpressinboxhubGetEmails 获取 expressinboxhub 邮件列表
 * token 格式: "cookie\ncsrf"
 */
func ExpressinboxhubGetEmails(email string, token string) ([]NormEmail, error) {
	parts := strings.SplitN(token, "\n", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return nil, fmt.Errorf("expressinboxhub: 无效的 session token")
	}
	cookie := parts[0]
	csrf := parts[1]

	body := "_token=" + csrf
	req, err := http.NewRequest("POST", expressinboxhubBase+"/messages", strings.NewReader(body))
	if err != nil {
		return nil, err
	}
	expressinboxhubAjaxHeaders(req)
	req.Header.Set("Cookie", cookie)
	req.Header.Set("X-CSRF-TOKEN", csrf)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "expressinboxhub getEmails"); err != nil {
		return nil, err
	}

	rawBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Mailbox  string                   `json:"mailbox"`
		Messages []map[string]interface{} `json:"messages"`
	}
	if err := json.Unmarshal(rawBytes, &data); err != nil {
		return nil, fmt.Errorf("expressinboxhub: 返回非 JSON: %s", string(rawBytes[:min(len(rawBytes), 120)]))
	}

	if data.Messages == nil || len(data.Messages) == 0 {
		return []NormEmail{}, nil
	}

	emails := make([]NormEmail, 0, len(data.Messages))
	for i, msg := range data.Messages {
		id := ""
		if v, ok := msg["id"]; ok {
			id = fmt.Sprintf("%v", v)
		} else {
			id = fmt.Sprintf("%d", i)
		}

		from, _ := msg["from"].(string)
		if from == "" {
			if v, ok := msg["sender"]; ok {
				from = fmt.Sprintf("%v", v)
			}
		}

		to, _ := msg["to"].(string)
		if to == "" {
			to = email
		}

		subject, _ := msg["subject"].(string)

		htmlContent, _ := msg["content"].(string)
		text, _ := msg["text"].(string)
		if text == "" {
			text, _ = msg["body"].(string)
		}

		date, _ := msg["receivedAt"].(string)
		if date == "" {
			date, _ = msg["date"].(string)
		}
		if date == "" {
			date, _ = msg["created_at"].(string)
		}

		flat := map[string]interface{}{
			"id":          id,
			"from":        from,
			"to":          to,
			"subject":     subject,
			"text":        text,
			"html":        htmlContent,
			"date":        date,
			"isRead":      false,
			"attachments": []any{},
		}
		emails = append(emails, NormalizeMap(flat, email))
	}
	return emails, nil
}

File: sdk/npm/src/providers/expressinboxhub.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'expressinboxhub';
const BASE = 'https://expressinboxhub.com';

/** CSRF token 提取正则 */
const CSRF_RE = /<meta\s+name="csrf-token"\s+content="([^"]+)"/i;

const DEFAULT_HEADERS: Record<string, string> = {
  Accept: 'application/json, text/plain, */*',
  'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
  'Cache-Control': 'no-cache',
  DNT: '1',
  Pragma: 'no-cache',
  Referer: `${BASE}/`,
  'sec-ch-ua': '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'sec-fetch-dest': 'empty',
  'sec-fetch-mode': 'cors',
  'sec-fetch-site': 'same-origin',
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
};

/** token 前缀，用于区分不同渠道的 token 格式 */
const TOK_PREFIX = 'eih1:';

/** 从 Set-Cookie 响应头中提取所有 cookie 行 */
function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === 'function') {
    return h.getSetCookie();
  }
  const one = headers.get('set-cookie');
  return one ? [one] : [];
}

/** 解析 cookie 字符串为键值映射 */
function parseCookieMap(hdr: string): Map<string, string> {
  const m = new Map<string, string>();
  for (const part of hdr.split(';')) {
    const p = part.trim();
    if (!p || !p.includes('=')) continue;
    const i = p.indexOf('=');
    const k = p.slice(0, i).trim();
    const v = p.slice(i + 1).trim();
    if (k) m.set(k, v);
  }
  return m;
}

/** 合并已有 cookie 与响应头中的新 cookie */
function mergeCookieHeader(prev: string, headers: Headers): string {
  const m = parseCookieMap(prev);
  for (const line of setCookieLines(headers)) {
    const first = line.split(';')[0] ?? '';
    const i = first.indexOf('=');
    if (i <= 0) continue;
    const k = first.slice(0, i).trim();
    const v = first.slice(i + 1).trim();
    if (k) m.set(k, v);
  }
  return [...m.entries()]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([k, v]) => `${k}=${v}`)
    .join('; ');
}

/** 将 CSRF token 与 session cookie 编码为 SDK 内部 token */
function encToken(csrf: string, cookie: string): string {
  const raw = JSON.stringify({ csrf, cookie });
  return TOK_PREFIX + Buffer.from(raw, 'utf8').toString('base64');
}

/** 解码 SDK 内部 token，还原 CSRF token 与 session cookie */
function decToken(tok: string): { csrf: string; cookie: string } {
  if (!tok.startsWith(TOK_PREFIX)) {
    throw new Error('expressinboxhub: invalid token');
  }
  let raw: string;
  try {
    raw = Buffer.from(tok.slice(TOK_PREFIX.length), 'base64').toString('utf8');
  } catch {
    throw new Error('expressinboxhub: invalid token');
  }
  const o = JSON.parse(raw) as { csrf?: string; cookie?: string };
  const csrf = (o.csrf ?? '').trim();
  const cookie = (o.cookie ?? '').trim();
  if (!csrf || !cookie) throw new Error('expressinboxhub: invalid token');
  return { csrf, cookie };
}

/**
 * 创建临时邮箱
 *
 * 流程：
 * 1. GET 首页获取 CSRF token 和 session cookie
 * 2. POST /messages 使用 CSRF token 创建邮箱
 *
 * @returns 包含邮箱地址和内部 token 的邮箱信息
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  /* 第一步：获取首页 HTML，提取 CSRF token 和 session cookie */
  const r1 = await fetchWithTimeout(BASE, {
    method: 'GET',
    headers: {
      ...DEFAULT_HEADERS,
      Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    },
  });
  if (!r1.ok) {
    throw new Error(`expressinboxhub: 获取首页失败 HTTP ${r1.status}`);
  }
  const html = await r1.text();
  let cookieHdr = mergeCookieHeader('', r1.headers);

  /* 从 HTML 中提取 CSRF token */
  const csrfMatch = CSRF_RE.exec(html);
  if (!csrfMatch?.[1]) {
    throw new Error('expressinboxhub: 未找到 CSRF token');
  }
  const csrfToken = csrfMatch[1];

  /* 第二步：POST /messages 创建邮箱 */
  const r2 = await fetchWithTimeout(`${BASE}/messages`, {
    method: 'POST',
    headers: {
      ...DEFAULT_HEADERS,
      'Content-Type': 'application/json',
      'X-CSRF-TOKEN': csrfToken,
      Cookie: cookieHdr,
    },
    body: JSON.stringify({ _token: csrfToken }),
  });
  if (!r2.ok) {
    throw new Error(`expressinboxhub: 创建邮箱失败 HTTP ${r2.status}`);
  }
  cookieHdr = mergeCookieHeader(cookieHdr, r2.headers);

  const data = (await r2.json()) as { mailbox?: string; messages?: unknown[] };
  if (!data.mailbox?.trim()) {
    throw new Error('expressinboxhub: 创建邮箱返回无效');
  }

  const email = data.mailbox.trim();
  const token = encToken(csrfToken, cookieHdr);

  return {
    channel: CHANNEL,
    email,
    token,
  };
}

/**
 * 获取邮件列表
 *
 * 使用创建邮箱时保存的 CSRF token 和 session cookie 进行请求
 *
 * @param token - SDK 内部 token（编码了 CSRF token 和 session cookie）
 * @param email - 邮箱地址
 * @returns 标准化后的邮件列表
 */
export async function getEmails(token: string, email: string): Promise<Email[]> {
  const { csrf, cookie } = decToken(token);

  const r = await fetchWithTimeout(`${BASE}/messages`, {
    method: 'POST',
    headers: {
      ...DEFAULT_HEADERS,
      'Content-Type': 'application/json',
      'X-CSRF-TOKEN': csrf,
      Cookie: cookie,
    },
    body: JSON.stringify({ _token: csrf }),
  });
  if (!r.ok) {
    throw new Error(`expressinboxhub: 获取邮件失败 HTTP ${r.status}`);
  }

  const data = (await r.json()) as { mailbox?: string; messages?: unknown[] };
  if (!Array.isArray(data.messages)) {
    return [];
  }

  return data.messages.map((raw) => normalizeEmail(raw, email));
}
